### PR TITLE
api: Add Golang module for APIs.

### PR DESCRIPTION
**Pull request by s3rj1k**
Add Golang module for APIs.

This change is beneficial to other projects that use rook API and want to keep import dependency minimal.
Importing API only reduces imported dependencies VS importing the whole rook.

### CODE REVIEW

**Comment by galexrt** [comment]
I'll review it in the coming hours. In itself, this change makes sense to make it easier for projects consuming the Rook APIs via Go.

**Comment by travisn** [comment]
Could we make this part of the [go.mod.update](https://github.com/rook/rook/blob/master/build/makelib/golang.mk#L167) make command? We need to automate it in some way.

**Comment by s3rj1k** [comment]
> Could we make this part of the [go.mod.update](https://github.com/rook/rook/blob/master/build/makelib/golang.mk#L167) make command? We need to automate it in some way.

Do you want this in this PR on a separate one?

**Comment by travisn** [comment]
As part of this PR sounds good.

I wonder if we could also make this work with the dependabot PRs such as #12157. Otherwise, we would need to open a separate PR every time the dependabot updates some dependencies. Maybe that's fine, but I'm wondering if we can avoid that.

**Comment by s3rj1k** [comment]
Regarding dependabot https://github.com/dependabot/dependabot-core/issues/1758 for inspiration.
Sound like it can be done.

**Comment by s3rj1k** [comment]
It seems that `go.mod.check` would be a better place for syncing, so it would always be synced and not directly tied to module updates.

PTAL

**Comment by travisn** [comment]
The mac build is failing in the CI, though I don't see it reporting a specific error.

When I build locally with this PR on my mac, I see this error:
```
=== syncing root modules with APIs modules
sed: 1: "pkg/apis/go.mod": extra characters at the end of p command
make: *** [go.mod.check] Error 1
```

**Comment by s3rj1k** [comment]
> The mac build is failing in the CI, though I don't see it reporting a specific error.

I guess it is failing on `-i` with sed, known issue for Macs.

What would be a more Mac way of replacing strings in file?

File: build/makelib/golang.mk [hunk 1]
```diff
@@ -170,7 +170,14 @@ go.mod.update:
 
 .PHONY: go.mod.check
 go.mod.check:
-	@echo === ensuring modules are tidied
+	@echo === syncing root modules with APIs modules
+	@cp -a go.mod pkg/apis/go.mod
+	@cp -a go.sum pkg/apis/go.sum
+	@sed -i 's|^module github.com/rook/rook|module github.com/rook/rook/pkg/apis|' pkg/apis/go.mod
```
Comment (subhamkrai) [inline build/makelib/golang.mk:176]: `sed -i` have an issue with Mac that's why we are seeing the build issue.

I did this and passed that error stage 
```git diff
diff --git a/build/makelib/golang.mk b/build/makelib/golang.mk
index a18c64ffc..3cb7de8c1 100644
--- a/build/makelib/golang.mk
+++ b/build/makelib/golang.mk
@@ -173,8 +173,8 @@ go.mod.check:
        @echo === syncing root modules with APIs modules
        @cp -a go.mod pkg/apis/go.mod
        @cp -a go.sum pkg/apis/go.sum
-       @sed -i 's|^module github.com/rook/rook|module github.com/rook/rook/pkg/apis|' pkg/apis/go.mod
-       @sed -i '\:^replace github.com/rook/rook/pkg/apis => ./pkg/apis:d' pkg/apis/go.mod
+       @sed -i '' 's|^module github.com/rook/rook|module github.com/rook/rook/pkg/apis|' pkg/apis/go.mod
+       @sed -i '' '\:^replace github.com/rook/rook/pkg/apis => ./pkg/apis:d' pkg/apis/go.mod
        @echo === ensuring APIs modules are tidied
        @(cd pkg/apis/; $(GOHOST) mod tidy -compat=1.17)
        @echo === ensuring root modules are tidied
```

**Review by subhamkrai — CHANGES_REQUESTED** [review]

File: build/makelib/golang.mk (hunk 1 above)
Comment (s3rj1k) [reply build/makelib/golang.mk]: This way it wound work on Linux

File: go.mod [hunk 2]
```diff
@@ -2,6 +2,8 @@ module github.com/rook/rook
 
 go 1.19
 
+replace github.com/rook/rook/pkg/apis => ./pkg/apis
```
Comment (Madhu-1) [inline go.mod:5]: I've noticed that many projects use `pkg/api` instead of `pkg/apis`. Is there a plan to continue using `api` instead of `apis`?

Comment (s3rj1k) [reply go.mod:5]: I would not want to change the project layout with this PR.

File: build/makelib/golang.mk (hunk 1 above)
Comment (s3rj1k) [reply build/makelib/golang.mk]: fixed for both Linux and Mac

File: build/makelib/golang.mk [hunk 3]
```diff
@@ -170,7 +170,13 @@ go.mod.update:
 
 .PHONY: go.mod.check
 go.mod.check:
-	@echo === ensuring modules are tidied
+	@echo === syncing root modules with APIs modules
+	@cp -a go.sum pkg/apis/go.sum
+	@cat go.mod | sed -e 's|^module github.com/rook/rook|module github.com/rook/rook/pkg/apis|' \
```
Comment (galexrt) [inline build/makelib/golang.mk:175]: `sed` is always fun on Mac, BSD vs GNU sed, I just remembered this script in the Rook repo which we use to workaround this situation: https://github.com/rook/rook/blob/7627a92319cd8797c4feceaad9d43ba9b3ad5a08/build/sed-in-place

File: build/makelib/golang.mk [hunk 4]
```diff
@@ -170,7 +170,13 @@ go.mod.update:
...
+	                  -e '\:^replace github.com/rook/rook/pkg/apis => ./pkg/apis:d' > pkg/apis/go.mod
+	@echo === ensuring APIs modules are tidied
+	@(cd pkg/apis/; $(GOHOST) mod tidy -compat=1.17)
```
Comment (travisn) [inline build/makelib/golang.mk:178]: Rook is building with go 1.19, do we still need this compat flag?

Comment (s3rj1k) [reply build/makelib/golang.mk:178]: It depends if rook whats to support go 1.17 if not I would keep the golang but change it's value to 1.19

Comment (travisn) [reply build/makelib/golang.mk:178]: Rook attempts to maintain compatibility with the two most recent golang versions, so 1.19 is the min version currently

Comment (s3rj1k) [reply build/makelib/golang.mk]: fixed

File: build/makelib/golang.mk (hunk 3 above)
Comment (s3rj1k) [reply build/makelib/golang.mk:175]: Mac is so unexpected :)

**Comment by travisn** [comment]
@s3rj1k The CI is failing on the latest changes

**Review by travisn — APPROVED** [review]
LGTM, let's see if there is impact on the dependabot in case we need a follow-up PR.

**Comment by s3rj1k** [comment]
@travisn Thanks.

**Pull request merged**